bootstrap/bootstrap.go

```diff
@@ -946,9 +951,63 @@ func hasGitSubmodules(sh *shell.Shell) bool {
 	return fileExists(filepath.Join(sh.Getwd(), ".gitmodules"))
 }
 
+func (b *Bootstrap) referenceRepository() (string, error) {
+	path := os.Getenv(`BUILDKITE_REPO_MIRROR_PATH`)
+	if path == "" {
+		path = filepath.Join(b.Config.ReposPath, dirForRepository(b.Repository))
+	}
+
+	// Create the base dir if it doesn't exist
+	if baseDir := filepath.Dir(path); !fileExists(baseDir) {
+		b.shell.Commentf("Creating \"%s\"", baseDir)
+		if err := os.MkdirAll(baseDir, 0777); err != nil {
+			return "", err
+		}
+	}
+
+	// Try and lock the repository dir to prevent concurrent clones
+	repoDirLock, err := b.shell.LockFile(path+".lock", time.Minute*5)
```

Comment:
Could this timeout please be surfaced? We have git repositories that take more than 5 minutes to clone.

We're _expecting_ the perforce repositories that we _will_ have to be multi-TB, which will _definitely_ require more than 5 minutes.

Reply:
This is configurable now!